modules/service/src/main/scala/lucuma/odb/service/ObservationWorkflowService.scala

```diff
@@ -279,22 +281,23 @@ object ObservationWorkflowService {
       private def cfpInfos(ids: List[CallForProposals.Id])(using Transaction[F]): F[Map[CallForProposals.Id, CfpInfo]] =
         NonEmptyList.fromList(ids).fold(Map.empty.pure[F]): nel =>
           val enc = cfp_id.nel(nel)
-          session 
+          session
             .stream(Statements.CfpInfos(enc))(nel, 1024)
+            // We get a row per instrument, so we need to group by CFP ID and fold the instruments
+            .fold(Map.empty[CallForProposals.Id, CfpInfo]) {
+              case (m, (cfp, oinst)) =>
+                m.updatedWith(cfp.cfpid) {
+                  case None    => cfp.addInstrument(oinst).some
+                  case Some(c) => c.addInstrument(oinst).some
+                }
+            }
```

Comment:
The `.fold` could come on the `CompileOps` result of `compile` and eliminate the `.last.map(_.getOrElse(Map.empty))` bit at the end.